### PR TITLE
Add success and fail callback of grpc sending reply

### PR DESCRIPTION
**Pull request by jiangzihao2009**
## What do these changes do?
At present, we send reply after handle the grpc request. However, the success and failure of sending reply are ignored, and handling the sending failure are impossible. Add success and fail callback in this pr.

## Linter

- [x] I've run `scripts/format.sh` to lint the changes in this PR.

### CODE REVIEW

File: src/ray/rpc/grpc_server.cc [hunk 1]
```diff
@@ -63,14 +63,21 @@ void GrpcServer::PollEventsFromCompletionQueue() {
...
+        // The rpc call has finished and can be deleted now.
         delete_call = true;
         break;
       default:
         RAY_LOG(FATAL) << "Shouldn't reach here.";
         break;
       }
+    } else {
+      // Code path for having sent reply to client and failed. PENDING call will be here
+      // when closing the grpc server.
```
Comment (raulchen) [inline src/ray/rpc/grpc_server.cc]: Consolidate this comment with the one on line 52-53

File: src/ray/rpc/server_call.h [hunk 2]
```diff
@@ -175,6 +194,12 @@ class ServerCallImpl : public ServerCall {
   /// The reply message.
   Reply reply_;
 
+  /// The callback when sending reply successes.
+  std::function<void()> send_reply_success_callback_ = nullptr;
+
+  /// The callback when sending reply fails.
+  std::function<void()> send_reply_fail_callback_ = nullptr;
```
Comment (raulchen) [inline src/ray/rpc/server_call.h]: nit, `send_reply_failure_callback_` (aligned with `success_callback`)

File: src/ray/core_worker/transport/raylet_transport.cc [hunk 3]
```diff
@@ -37,14 +37,14 @@ CoreWorkerRayletTaskReceiver::CoreWorkerRayletTaskReceiver(
...
-    rpc::RequestDoneCallback done_callback) {
+    rpc::SendReplyCallback send_reply_callback) {
   const std::string &task_message = request.task_spec();
   const raylet::Task task(*flatbuffers::GetRoot<protocol::Task>(
       reinterpret_cast<const uint8_t *>(task_message.data())));
   const auto &spec = task.GetTaskSpecification();
 
   auto status = task_handler_(spec);
-  done_callback(status);
+  send_reply_callback(status, nullptr, nullptr);
```
Comment (raulchen) [inline src/ray/core_worker/transport/raylet_transport.cc]: can we use `nullptr` as default value?

File: src/ray/rpc/server_call.h [hunk 4]
```diff
@@ -123,30 +127,45 @@ class ServerCallImpl : public ServerCall {
...
 
-  /// Tell gRPC to finish this request.
-  void Finish(Status status) override {
+  void SendReplySuccessCallback() {
+    if (send_reply_success_callback_) {
+      send_reply_success_callback_();
+    }
+  }
+
+  void SendReplyFailCallback() {
```
Comment (raulchen) [inline src/ray/rpc/server_call.h]: `OnReplySent` and `OnReplyFailed`?
Also, document these 2 methods.

File: src/ray/rpc/server_call.h [hunk 5]
```diff
@@ -11,7 +11,8 @@ namespace rpc {
 
 /// Represents the callback function to be called when a `ServiceHandler` finishes
 /// handling a request.
-using RequestDoneCallback = std::function<void(Status)>;
+using SendReplyCallback =
+    std::function<void(Status, std::function<void()>, std::function<void()>)>;
```
Comment (raulchen) [inline src/ray/rpc/server_call.h]: Add argument names and document the args.

**Review by raulchen — COMMENTED** [review]
Looks good overall, left a few small comments.

File: src/ray/core_worker/transport/raylet_transport.cc (hunk 3 above)
Comment (jiangzihao2009) [reply src/ray/core_worker/transport/raylet_transport.cc]: It seems that default value is not allowed in lambda function in c++11

Comment (raulchen) [reply src/ray/core_worker/transport/raylet_transport.cc]: It's allowed.
```
    auto f = [](int a = 1) {
        std::cout << a << std::endl;
    };
    f();
```
prints `1`.

File: src/ray/rpc/grpc_server.cc [hunk 6]
```diff
@@ -49,8 +49,10 @@ void GrpcServer::PollEventsFromCompletionQueue() {
   // Keep reading events from the `CompletionQueue` until it's shutdown.
   while (cq_->Next(&tag, &ok)) {
     auto *server_call = static_cast<ServerCall *>(tag);
-    // `ok == false` indicates that the server has been shut down.
-    // We should delete the call object in this case.
+    // `ok == false` will occur in two situations:
+    // First, the server has been shut down, the server call's status is PENDING
+    // Second, server has sent reply to client and failed, the server call's status is
+    // SENDING_REPLY
```
Comment (raulchen) [inline src/ray/rpc/grpc_server.cc]: put this comment in the `else` block`?

File: src/ray/rpc/server_call.h [hunk 7]
```diff
@@ -11,8 +11,11 @@ namespace rpc {
 
 /// Represents the callback function to be called when a `ServiceHandler` finishes
 /// handling a request.
-using SendReplyCallback =
-    std::function<void(Status, std::function<void()>, std::function<void()>)>;
+/// \param status The status would be returned to client.
+/// \param success Success callback which will be invoked in function OnReplySent.
```
Comment (raulchen) [inline src/ray/rpc/server_call.h]: `... which will be invoked when the reply is successfully sent to the client.`

**Review by raulchen — APPROVED** [review]

**Comment by raulchen** [comment]
[this CI failure](https://travis-ci.com/ray-project/ray/jobs/213996058) looks related. 
And linter failed as well.

**Pull request merged**